### PR TITLE
To promise

### PR DESCRIPTION
**Pull request by haoxins**

### CODE REVIEW

**Comment by haoxins** [comment]
the first step to make `chan` work with `ES7 async/await` >_<

**Comment by haoxins** [comment]
cc @brentburg

**Comment by brentropy** [comment]
Sorry, I have been slow to get to this. I will take a look. A while back a started a big overhaul that was going to be released as `chan 1.0` that included migrating from thunks to promises, but I got busy and never finished. I want to compare some of your changes to mine, but I am probably good with going forward with your changes, since I don't have the time to dedicate to it right now. Thanks @coderhaoxin.

**Comment by haoxins** [comment]
@brentburg Anything else I can do to help the migration?

Can you give a roadmap or some thoughts?

I want to update `chan` in [cojs/busboy](https://github.com/cojs/busboy), and also some private projects (that will migrate from `co` to `async&await`).

**Comment by haoxins** [comment]
@brentburg Can we merge this first.

So, I can update chan in co-busboy, see 

https://github.com/cojs/busboy/issues/33
https://github.com/cojs/busboy/issues/30

File: README.md [hunk 1]
```diff
@@ -37,30 +38,26 @@ typeof ch // -> 'function'
...
 })
 ```
 
 ### Receiving values from the channel
 
-Values are removed from the channel by calling it with a node-style callback as
-this first argument. When a value is available on the channel the callback is
-called with the value or error. In this case the channel itself can also be a
-thunk.
+Values are removed from the channel by calling it with a Promise function chain.
```
Comment (juliangruber) [inline README.md]: I'd reword this to something like `Values are taken off the channel by resolving the channel promise repeatedly`. You don't have to chain to get values.

**Comment by haoxins** [comment]
ping @brentburg

File: README.md [hunk 2]
```diff
@@ -37,30 +38,26 @@ typeof ch // -> 'function'
...
-Values are removed from the channel by calling it with a node-style callback as
-this first argument. When a value is available on the channel the callback is
-called with the value or error. In this case the channel itself can also be a
-thunk.
+Values are taken off the channel by resolving the channel promise repeatedly.
 
 ```js
-ch(function (err, val) {
-  // called when there is a value or error on the channel
+ch.then(function (val) {
```
Comment (brentropy) [inline README.md:57]: This doesn't seem to be implemented. The channel is a function does not have a then method.

**Comment by brentropy** [comment]
Sorry, I have gone so long without doing anything with this. There seem to be some things missing at this point (i.e. the channel function isn't thenable) and looking at the test I am still seeing a lot of thunk stuff. I have a local branch I could push up (on my old laptop, so I can't right now) that was a deeper refactor to make everything promise friendly, but I still needed to rework the tests and probably fix some bugs.

If you want to push forward with your branch and fill in the gaps, that works for me, otherwise we can see what needs to be done to go forward with my branch.

**Comment by haoxins** [comment]
> the channel function isn't thenable, and looking at the test I am still seeing a lot of thunk stuff

Yea! This is just the first step to Promise :)

>  I have a local branch I could push up (on my old laptop, so I can't right now) that was a deeper refactor to make everything promise friendly

👍 @brentburg can you push this to github? So, maybe I can give some help.

**Comment by haoxins** [comment]
> If you want to push forward with your branch and fill in the gaps, that works for me, otherwise we can see what needs to be done to go forward with my branch.

I don't care `my branch` or `your branch`. Just want to finish that (to Promise) in a short time.

**Comment by brentropy** [comment]
@coderhaoxin see #18. It is a v1 branch and has other changes. Let me know what you think. I did a quick to-do list of what I think still needs to be done. I can try to spend some time on it too, but if you are interested in working on that branch that would be great.

**Comment by haoxins** [comment]
closed by #18

**Comment by haoxins** [comment]
@brentburg I'll check that this weekend!
